### PR TITLE
Feature/gpg

### PR DESCRIPTION
**Pull request by weatherhead99**
- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

This is a possible implementation of the feature request for GPG signature validation #1959 
(and a re-opening of previous PR after re-factoring and more work)

This adds a dependency on python-gnupg (http://pythonhosted.org/python-gnupg/index.html), which is pure python, but obviously requires a system gpg binary to work. This is the fork which doesn't require the problematic psutil module.

At present, the implementation detects this, and then throws on asking to verify a signature if python-gnupg didn't find a system gpg binary.

This also adds an extra install into the CI builds, not sure what the implications of that are.

### CODE REVIEW

**Comment by memsharded** [comment]
Hi @weatherhead99 

Thanks for your continued effort on this.

Unfortunately, I am seeing that python-gnupg still has a transitive dependency to ``psutil``, which was the original offender (breaking installation in some systems, much before conan python code could conditionally import anything).

I think this functionality should wait to be incorporated to conan until some plugin system enables conditional installation of conan extensions/components/plugins.

Please @lasote feedback.

**Comment by weatherhead99** [comment]
Hi @memsharded ,
are you sure you're not mistaken with the psutil dependency? Note the packages python-gnupg and gnupg are different. I am using this fork (version 0.4.1) http://pythonhosted.org/python-gnupg/index.html 
which as far as I can see uses only subprocess (it is a single python file).

I think perhaps you are still looking at this fork: https://github.com/isislovecruft/python-gnupg (version 2. something)



In any case, for verifying files we don't require all the functionality of python-gnupg. If it's still unacceptable, I am happy to write a subprocess wrapper that only works on files for this purpose - I do think GPG sig validation is an important feature, since many of the "old style" open source libraries (i.e. the ones that actually post tarballs on websites) are going to provide GPG validation as the main method to verify downloads.

**Comment by memsharded** [comment]
Yes, certainly, I was looking at: https://github.com/isislovecruft/python-gnupg/blob/master/requirements.txt.

Sorry for the mistake, I will have another look, thanks!

**Comment by memsharded** [comment]
I have had another look to this PR and I am very sorry to say that I have still some concerns:

- CI is broken for Windows and OSX. An evidence that it is not that simple to install the gpg tool in them. It would be necessary that everything is green.
- I am a bit uncomfortable to adding yet another thing to install for conan contributors running the test suite, and that would be different in every OS.
- What is the process for getting the keys? Putting them in a folder in user machine, right? Where to get these keys? Who is responsible of installing these keys, the users? From where? For every package that requires some gpg signing?
- It is some code that, at least me, is difficult to know what it is doing. E.g.: 
```
signed = _gpg.sign_file(f, keyid=FINGERPRINT[-8:],
+                                passphrase=PASSPHRASE, detach=True,
+                                binary=False,
```
Why the last 8 chars of the fingerprint? What is detach? Of course this is not inherent to the PR, but to the library used. But still something that will require a big effort on our side to maintain and support in the future.

The 1.2 release is next week, I don't think this feature is ready to be released yet, so it will most likely be postponed to a later release. Also, I need to understand better the scope and flow for this utility. If they work with signatures that could be embedded in the recipe, fine, but using an independent key file, I don't get it. Is it intended to work for packages in conan-center? 

cc/ @SSE4 who opened the issue for this feature.

Many thanks!

**Comment by weatherhead99** [comment]
Thanks for looking at it :-)

as to the CI comments - I believe that gpg is not installed in the jenkins instances. I was able to add it to the Travis  & Appveyor scripts, but I see those aren't being run on PRs at the moment?  Anyway, those failures at the moment are failing (so far as I can see from a first look) in the expected manner when gpg is not installed on the system. I don't know how to tell jenkins to install these utilities. 
We can very easily bypass these tests in the case that gpg isn't on the user's system...

The pubkey fingerprints for software projects are generally either published on the projects' website, or available from a GPG keyserver. For example, in the case of all the big GNU projects (e.g. GSL: https://www.gnu.org/software/gsl/, the key is found here: https://www.gnu.org/software/gsl/key/gsl_key.txt ).  There is no need to "install" the keys in the system gpg, as this implementation uses a separate GPG keyring to avoid pollution of the users regular keyring. You can pass it either a keyid (which is the last 8 characters of the fingerprint), or a filename. It will retrieve the key from a keyserver in the former case, open the file in the latter case. 

Every package that is GPG signed will of course be signed by the pubkey of that particular developer. 

"detach" refers to whether the signature is detached (i.e. software.tar.gz + software.tar.gz.sig), or is integrated in the overall file (e.g. software.tar.gz.asc). I agree that python-gnupg is a bit of a nightmare to use.  Most software tends to use detached plaintext signatures, a notable exception for example is apache, which uses detached binary signatures:  https://www.apache.org/dyn/closer.cgi#verify
I don't believe I've ever seen a software which used non-detached signatures, though the case this might be useful is e.g. for a software project which provides a file which tells you what the MD5 / SHA1 of the download file is, then signs _that_ file in a non-detached way (several linux distributions do this for example, I know you wouldn't package one using conan!)

As stated before, this would massively help my personal use and confidence in conan, I'm willing to try and wrap the bits of gpg we need a bit more nicely if that would help acceptance.

Overall, I think that for "old fashioned" software projects that still host their released tarballs on websites, GPG sigs are one of the major ways to verify downloads (the other being MD5 / SHA1 etc). And so, if conan wants to support those softwares properly (including downloading source on demand) it kind of has to support gpg verification.

**Comment by memsharded** [comment]
Thanks very much for all the feedback.

Yes, that is true, we need to have a look to the CI servers to install gpg, as I think it is not possible directly from jenkins scripts.

File: conans/requirements.txt [hunk 1]
```diff
@@ -11,3 +11,4 @@ pylint>=1.8.1, <1.9.0
 future==0.16.0
 pygments>=2.0, <3.0
 astroid>=1.6, <1.7
+python-gnupg>=0.4.1
```
Comment (memsharded) [inline conans/requirements.txt]: Why requiring this, instead of the https://pypi.python.org/pypi/gnupg (2.3.1)?

Comment (memsharded) [reply conans/requirements.txt]: Well, I answer myself: License: GPLv3+ is incompatible with MIT

File: conans/client/tools/net.py [hunk 2]
```diff
@@ -3,13 +3,13 @@
...
 from conans.client.rest.uploader_downloader import Downloader
-from conans.client.tools.files import unzip, check_md5, check_sha1, check_sha256
+from conans.client.tools.files import unzip, check_md5, check_sha1, check_sha256, verify_gpg_sig
 from conans.errors import ConanException
 
 _global_requester = None
 
 
-def get(url, md5='', sha1='', sha256='', destination="."):
+def get(url, md5='', sha1='', sha256='', gpg_signature='', gpg_pubkey='', destination="."):
```
Comment (memsharded) [inline conans/client/tools/net.py]: I am a bit confused about the input parameters, and the underlying verify_gpg_sig() function. This assumes:
- always using "keys.gnupg.net" as server?
- If using a file for pubkey, what is the path of the file? Relative to the current folder? The file has been ``exports`` with the recipe?

File: conans/test/functional/gpg_verify_test.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    def tearDown(self):
+        map(os.remove, self._data_files)
+
+    def testCorrectVerifyFile(self):
+        verify_gpg_sig(data_file=self._data_file, pubkey=GPG_PUBLIC_KEY,
+                       sig_file=self._sig_file)
+
+        #if we got to here without throwing, verification completed
+
+    def testIncorrectSignatureVerifyFile(self):
```
Comment (memsharded) [inline conans/test/functional/gpg_verify_test.py]: Need to use ``test_incorrect_...`` pep8 names.

Also make sure your code is linted, it seems you have no checker.

File: conans/requirements.txt (hunk 1 above)
Comment (weatherhead99) [reply conans/requirements.txt]: in addition, that other fork is the one that needs psutil

File: conans/client/tools/net.py (hunk 2 above)
Comment (weatherhead99) [reply conans/client/tools/net.py]: I mean, we could allow the user to supply their own keyserver for completenes, but the major ones are all synced. 

I would envisage the public key file is included in the recipe and exported.

File: conans/test/functional/gpg_verify_test.py (hunk 3 above)
Comment (weatherhead99) [reply conans/test/functional/gpg_verify_test.py]: I do have a checker, but probably not set high enough...

**Comment by weatherhead99** [comment]
as to integration testing, is there any existing test that exercises tools.get() without network access that you could point me to as a starting point? We certainly could sign some little file (or generate and sign, that's how the GPG functional tests work) and then shim it into tools.get.

**Comment by memsharded** [comment]
Update: this PR is pending work to be merged here: https://github.com/weatherhead99/conan/pull/5

**Comment by weatherhead99** [comment]
sorry I have been MIA (real world work stuff!). This is still ongoing for a few weeks, but I have got round to merging your PR.

**Comment by jgsogo** [comment]
Now that hooks are a reality and we are getting some traction and feedback about them, I think that this functionality could be implemented using one of them, don't you think so? We wrote a [blogpost about hooks](https://blog.conan.io/2019/02/28/Extending-conan-with-hooks.html) that could serve as an entry point before approaching [the docs](https://docs.conan.io/en/latest/extending/hooks.html#hooks). Using hooks we can add functionality that the user can easily opt-in and we don't need it to work for every platform we support, also there won't be any issue with the licenses as we are not providing the libraries along with Conan.

If you think it is the way to go, I encourage you to implement the hook youself; if not, we can [open an issue there](https://github.com/conan-io/hooks) and someone will take care of it.

Thanks!

**Comment by memsharded** [comment]
During all this time, it seems that there was not enough user demand for this. Moving the ConanCenter to a centralized CI seems to offer enough guarantees, and company and enterprise users are typically building their own binaries anyway.

Signing of packages is now probably easily implemented as hooks, we will try to build an example proof of concept after 2.0, but at this moment this can be closed. Thanks anyway for all the effort that was done that time ago!

**Pull request closed without merging**